packages/utils/src/dal/mikro-orm/big-number-field.ts

```diff
@@ -1,78 +1,46 @@
 import { BigNumber } from "../../totals/big-number"
-
-const bigNumberFields = new WeakMap<
-  object,
-  { prop: string; options: { nullable?: boolean } }[]
->()
-
-export function BigNumberField(options: { nullable?: boolean } = {}) {
-  return function (target: any, prop: string) {
-    const entity = target.constructor
-    if (!bigNumberFields.has(entity)) {
-      bigNumberFields.set(entity, [])
-    }
-
-    if (prop.startsWith("raw_")) {
-      const suggestedPropName = prop.replace("raw_", "")
-      throw new Error(
-        `BigNumberField decorator has to be used on the property "${suggestedPropName}" and ${prop} typed as BigNumberRawValue.`
-      )
-    }
-
-    bigNumberFields.get(entity)?.push({ prop, options })
-
-    if (!entity.prototype.__bigNumberInitialized) {
-      entity.prototype.__bigNumberInitialized = true
-
-      registerGlobalHook(entity)
-    }
-  }
-}
-
-function registerGlobalHook(entity: any) {
-  const originalOnInit = entity.prototype.onInit
-  const originalOnCreate = entity.prototype.onCreate
-  const originalOnUpdate = entity.prototype.onUpdate
-
-  entity.prototype.onInit = function (...args: any[]) {
-    initializeBigNumberFields(this)
-
-    if (originalOnInit) {
-      originalOnInit.apply(this, args)
-    }
-  }
-
-  entity.prototype.onCreate = function (...args: any[]) {
-    initializeBigNumberFields(this)
-
-    if (originalOnCreate) {
-      originalOnCreate.apply(this, args)
-    }
-  }
-
-  entity.prototype.onUpdate = function (...args: any[]) {
-    initializeBigNumberFields(this)
-
-    if (originalOnUpdate) {
-      originalOnUpdate.apply(this, args)
-    }
-  }
-}
-
-function initializeBigNumberFields(entity: any) {
-  const fields = bigNumberFields.get(entity.constructor) ?? []
-
-  for (const field of fields) {
-    const { prop, options } = field
-
-    const rawValue = entity[`raw_${prop}`]
-    const value = entity[prop]
-    if (options.nullable && rawValue === null && value === null) {
-      return
-    }
-
-    const val = new BigNumber(rawValue ?? value)
-    entity[prop] = val.numeric
-    entity[`raw_${prop}`] = val.raw
+import { Property } from "@mikro-orm/core"
+import { isDefined } from "../../common"
+import { BigNumberInput } from "@medusajs/types"
+
+export function MikroOrmBigNumberProperty(
+  options: Parameters<typeof Property>[0] & {
+    rawColumnName?: string
+  } = {}
+) {
+  return function (target: any, columnName: string) {
+    const targetColumn = columnName + "_"
+    const rawColumnName = options.rawColumnName ?? `raw_${columnName}`
+
+    Object.defineProperty(target, columnName, {
+      get() {
+        if (!isDefined(this[rawColumnName]) && !isDefined(this[targetColumn])) {
+          return null
+        }
+
+        return this[targetColumn] instanceof BigNumber
```

Comment:
question: aren't we setting `this[targetColumn]` on L:28 as a number so this check will never pass i.e. we will always call BigNumber constructor

Reply:
should be all good man 👍